src/task_impl/std/mod.rs

```diff
@@ -651,3 +632,207 @@ impl<T> From<Arc<T>> for NotifyHandle
+/// executor. Crates such as `tokio-core`, for example, have a method to
+/// install the `Core` as a local executor for the duration of a closure.
+/// Crates like `rayon`, however, will automatically take care of this
+/// trait.
+pub trait Wait {
+    /// Returns an identifier to pass to the `poll_*_notify` function that's
+    /// about to be used.
+    ///
+    /// The identifier is called each time before `poll_*_notify` is called and
+    /// is immediately passed to the relevant function. Most waiters may not end
+    /// up using this and will also ignore the id passed to their
+    /// `NotifyHandle`.
+    fn notify_id(&self) -> usize;
+
+    /// Extract's a `NotifyHandle` to unblock this waiter.
+    ///
+    /// The purpose of the `wait_*`-style functions of this crate is to block
+    /// the calling context until a value becomes available. The `wait` method
+    /// below will be invoked to actually perform the blocking operation, but
+    /// we've also got to know when to wake up!
+    ///
+    /// This function returns a `NotifyHandle` suitable for use in a `Task` that
+    /// will be handed out as part of `task::current`. If a future returns
+    /// `NotReady` then we'll call `wait` below to block until the future is
+    /// ready. The future will internally use the `NotifyHandle` returned here
+    /// to indicate that it's ready to make progress.
+    ///
+    /// It is expected that the handle returned from this function can be used
+    /// to unblock calls to `wait` below.
+    fn notify_handle(&self) -> NotifyHandle;
+
+    /// Blocks execution of the current thread until this instance's
+    /// `NotifyHandle` is notified.
+    ///
+    /// This crate implements the various `wait` methods it provides with this
+    /// function. When an object is polled and determined to be not ready this
+    /// function is invoked to block the current thread. This may literally
+    /// block the thread immediately, or it may do other "useful work" in the
+    /// meantime, such as running an event loop.
+    ///
+    /// The crucial detail of this method is that it does not return until
+    /// the object being polled is ready to get polled again. That is defined
+    /// as when the `NotifyHandle` returned through the `notify` method above
+    /// is notified. Once a notification is received implementors of this trait
+    /// should ensure that this method returns shortly thereafter.
+    ///
+    /// Note that it is permissible for this function to have spurious wakeups.
+    /// If this function returns without the above handle being notified, that's
+    /// ok.
+    ///
+    /// # Panics
+    ///
+    /// It is expected that this method will not panic. This method may be
+    /// called recursively due to multiple invocations of `Future::wait` on the
+    /// stack.  In other words its expected for executors to handle
+    /// waits-in-waits correctly.
```

Comment:
@tanriol you are very correct. If nested calls to wait are done, then there are no guarantees as to when `wait` returns. In fact, there could be cases where `wait` never returns (because someone called `wait` on `tcp_listener.incoming().for_each(...)`.

This is definitely a significant hazard and great care should be taken with this feature.